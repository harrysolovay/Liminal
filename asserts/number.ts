import { assert } from "../util/mod.ts"

export function min(value: number, min: number) {
  assert(value >= min, `Must be gte ${min}; received ${value}.`)
}

export function max(value: number, max: number) {
  assert(value <= max, `Must be lte ${max}; received ${value}.`)
}
