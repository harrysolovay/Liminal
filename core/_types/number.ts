import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export const number: Type<number> = declareType(() => number)
