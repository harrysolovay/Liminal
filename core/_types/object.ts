import type { Type } from "../Type.ts"
import { declareType } from "./_declareType.ts"

export function object<F extends Record<string, Type>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }> {
  return declareType(() => object, [fields])
}
