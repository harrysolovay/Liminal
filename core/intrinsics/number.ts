import { declareType } from "../declareType.ts"
import type { Type } from "../Type.ts"

export const number: Type<number> = declareType({
  type: "number",
  self() {
    return number
  },
})
