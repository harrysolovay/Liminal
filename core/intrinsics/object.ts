import { Rune } from "../Rune.ts"
import { consumeType } from "./_common.ts"

export function object<F extends Record<string, Rune>>(
  fields: F,
): Rune<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["E"]> {
  return Rune({
    kind: "object",
    self: () => object,
    args: [fields],
    consume: consumeType,
  })
}
