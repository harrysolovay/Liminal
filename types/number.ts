import { declareType, type Type } from "../core/mod.ts"

export const number: Type<number> = declareType({
  name: "number",
  source: {
    getType: () => number,
  },
})
