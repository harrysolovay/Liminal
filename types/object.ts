import { type AnyType, declareType, type Type } from "../core/mod.ts"

export function object<F extends Record<string, AnyType>>(
  fields: F,
): Type<{ [K in keyof F]: F[K]["T"] }, F[keyof F]["P"]> {
  const entries = Object.entries(fields)
  return declareType({
    name: "object",
    source: {
      factory: object,
      args: [fields],
    },
    visitValue: (value, visit) => {
      const visited = Object.fromEntries(
        entries.map(([key, type]) => [
          key,
          visit(
            value[key],
            type,
            () => (value) => {
              visited[key] = value
            },
            {
              formatValuePath: (leading) => `${leading}.${key}`,
              formatTypePath: (leading) => `${leading}.${key}`,
            },
          ),
        ]),
      )
      return visited as never
    },
  })
}
