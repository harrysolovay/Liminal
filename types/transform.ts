import { Ty } from "./Ty.ts"

export function transform<T, P extends keyof any, R extends boolean, I>(
  ty: Ty<I, P, R>,
  transform: (value: I) => T,
): Ty<T, P, R> {
  return Ty<T, P, R, I>((ref) => ref(ty), ty[""].root, (value) => transform(value))
}
