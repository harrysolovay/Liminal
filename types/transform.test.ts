import { assertTypeSnapshot } from "testing"
import { number } from "./number.ts"
import { object } from "./object.ts"
import { transform } from "./transform.ts"

Deno.test("transform", async (t) => {
  await assertTypeSnapshot(
    t,
    transform(
      "Coordinates",
      object({
        latitude: number,
        longitude: number,
      }),
      ({ latitude, longitude }) => [latitude, longitude],
    ),
  )
})
