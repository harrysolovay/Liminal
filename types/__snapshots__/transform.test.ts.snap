export const snapshot = {};

snapshot[`transform 1`] = `
{
  additionalProperties: false,
  properties: {
    latitude: {
      type: "number",
    },
    longitude: {
      type: "number",
    },
  },
  required: [
    "latitude",
    "longitude",
  ],
  type: "object",
}
`;

snapshot[`transform 2`] = `
{
  additionalProperties: false,
  description: "One.",
  properties: {
    latitude: {
      type: "number",
    },
    longitude: {
      type: "number",
    },
  },
  required: [
    "latitude",
    "longitude",
  ],
  type: "object",
}
`;

snapshot[`transform 3`] = `
{
  additionalProperties: false,
  description: "One. Two.",
  properties: {
    latitude: {
      type: "number",
    },
    longitude: {
      type: "number",
    },
  },
  required: [
    "latitude",
    "longitude",
  ],
  type: "object",
}
`;
